### ISSUE TITLE
npm postinstall fails behind an HTTPS proxy

### ISSUE THREAD

**Issue by Toilal** [cypress-io/cypress#1401]
### Current behavior:

I'm running behind an HTTPS corporate proxy configured with a self-signed certificate. This requires to setup the self-signed certificate with `NODE_EXTRA_CA_CERTS` environment variable.

`yarn install cypress` gives the following error (same with npm)

```
> cypress@2.1.0 postinstall /app/oison-frontend/node_modules/cypress
> node index.js --exec install

Installing Cypress (version: 2.1.0)

 ✖  Downloading Cypress
   → Cypress Version: 2.1.0
    Unzipping Cypress
    Finishing Installation
The Cypress App could not be downloaded.

Please check network connectivity and try again:
----------

URL: https://download.cypress.io/desktop/2.1.0?platform=linux&arch=x64
Error: self signed certificate in certificate chain
----------

Platform: linux (Debian - 8.10)
Cypress Version: 2.1.0
```

It seems `NODE_EXTRA_CA_CERTS` is not honored by the post installation script downloading cypress binary.

### Desired behavior:

It should work

### How to reproduce:

Come to my company and have fun configuring this proxy sh*t.

### Additional Info (images, stack traces, etc)

It was working yesterday, so I think something has changed in the package of `2.1.0` release.

- Operating System: Ubuntu
- Cypress Version: >=2.1
- Browser Version:

**Comment by gkemp94** [comment]
I had this problem a few days ago and found that the best workaround was to download the zip file directly from the site then type the following command before attempting npm install. This should then install the file from your zip file. Hope that helps. 

`CYPRESS_BINARY_VERSION=/local/path/to/cypress.zip `

**Comment by perceboisElisa** [comment]
Hi, 
Download the .zip is the solution when downloading cypress with npm command is blocked by your corporate proxy.

But then could you run your app within cypress? 

...because for me, I can launch cypress.exe but it is blocked due to your corporate proxy and i have currently no idea to solve this problem :-(

Regards,

**Comment by Toilal** [comment]
I don't really use Cypress for now, it's just embedded in the default VueJS boilerplate for E2E tests.

**Comment by zerdos** [comment]
Try this environment variable:
export NODE_TLS_REJECT_UNAUTHORIZED=0

**Comment by davidsteinsland** [comment]
Cypress should (at least) honor the following env vars, which is the de facto way of configuring proxy:
- HTTP_PROXY / http_proxy
- HTTPS_PROXY / https_proxy

Also, because it's node:
- NODE_EXTRA_CA_CERTS

And perhaps also the following npm configs:
* `cafile` 
* `https-proxy` 
* `proxy`

Stuff like setting `NODE_TLS_REJECT_UNAUTHORIZED=0` or similar is out of the question, and defeats the whole purpose of secure connections.

**Comment by jdhines** [comment]
@gkemp94's answer solved it for me.

**Comment by ggarcia92** [comment]
@gkemp94's answer also solved it for me.

**Comment by brian-mann** [comment]
@davidsteinsland Cypress should already support the `HTTP_PROXY` and `HTTPS_PROXY` env vars since it uses `request` under the hood to download the binary...

**Comment by brian-mann** [comment]
@jdhines and @ggarcia92 can you provide that as an env var during `npm install` or export it to your profile ahead of time?

**Comment by uriannrima** [comment]
@brian-mann I was having the same issue of ECONNREFUSED for being behind corporate proxy, had already configured `PROXY` and `HTTP-PROXY` on npm config list, but it still ocurred. Configured `HTTP_PROXY` and `HTTPS_PROXY` on system environments and it finally downloaded successfully.

(It was needed to use @vue/cli template)

**Comment by jennifer-shehane** [comment]
@uriannrima awesome! There is some workaround documentation on this exact thing in our Epic proxy issue here: https://github.com/cypress-io/cypress/issues/1469

**Comment by zerdos** [comment]
so, if you have man in the middle corporate proxy, you can add the extra certificate as an environment variable.

It's solved our problem.

example:

NODE_EXTRA_CA_CERTS=./ssl/mitn.pem yarn

**Comment by squarism** [comment]
@zerdos My hero of the week.  I was doing the binary env variable for this project but that little extra setting (which I was diving into a bunch of sources to find) worked and will likely solve future problems for me.

**Comment by dschulten** [comment]
In a Jenkinsfile for a Jenkins with Credentials Plugin and [Credentials Binding Plugin](https://jenkins.io/doc/pipeline/steps/credentials-binding/) installed and a username-password credential definition having the id PROXY_USERPASS the following configures the proxy for cypress postinstall and does not print the credentials into the log:

    withCredentials([usernameColonPassword(credentialsId: 'PROXY_USERPASS', variable: 'USERPASS')]) 
    {
      sh '''
      set +x
      sh 'HTTPS_PROXY=http://<user>:<url-encoded-passwd>@corporate.proxy:<port> npm install'
      '''
    }

Also, caching of the cypress binary works well, it only downloads once.

**Comment by madelson** [comment]
Would be great to see this fixed

**Comment by jmclocklin** [comment]
This issue is blocking us from using Cypress - a fix would be super useful.

**Comment by jennifer-shehane** [comment]
@flotwig Another potential proxy issue. Is this not fixed in 3.3.0 PR? https://github.com/cypress-io/cypress/pull/3531

**Comment by flotwig** [comment]
@jennifer-shehane This issue should be fixed by #3531, since now we have support for proxies and `NODE_EXTRA_CA_CERTS` is automatically dealt with by node in `tls`, so we don't have to do anything special.

**Comment by jennifer-shehane** [comment]
The code for this is done in https://github.com/cypress-io/cypress/pull/3531, but this has yet to be released. We'll update this issue and reference the [changelog](https://on.cypress.io/changelog) when it's released.

**Comment by jinyangzhen** [comment]
> I had this problem a few days ago and found that the best workaround was to download the zip file directly from the site then type the following command before attempting npm install. This should then install the file from your zip file. Hope that helps.
> 
> `CYPRESS_BINARY_VERSION=/local/path/to/cypress.zip `

renamed to 
export CYPRESS_INSTALL_BINARY=/home/b_bis/cypress.zip

**Comment by JMclovin** [comment]
@jennifer-shehane Any time frame when this fix might be release as this issue is stopping us from using cypress in our environments

**Comment by ravikiranr26** [comment]
Hi Sir/Mam,
CYPRESS_INSTALL_BINARY=C:\Users\xyz\work\work\cypress.zip npm install cypress
Even doing a local install to bypass.

Log:
node index.js --exec install

‼ Warning: Forcing a binary version different than the default.

  The CLI expected to install version: 3.2.0

  Instead we will install version: C:Usersrramarajuworkworkcypress.zip

  These versions may not work properly together.

Installing Cypress (version: C:Usersrramarajuworkworkcypress.zip)

[10:29:58]  Downloading Cypress     [started]
[10:29:59]  Downloading Cypress     [failed]
[10:29:59] → The Cypress App could not be downloaded.

Please check network connectivity and try again:
----------

URL: https://download.cypress.io/desktop/C:Usersrramarajuworkworkcypress.zip?platform=win32&arch=x64
Error: self signed certificate in certificate chain
----------

Platform: win32 (10.0.16299)
Cypress Version: 3.2.0
The Cypress App could not be downloaded.

Please check network connectivity and try again:

Regards.
Ravi

**Comment by jennifer-shehane** [comment]
We're working on patching up fixes to release this as fast as we can. The introduction of the proxy code brought up some regressions we needed to address before releasing.

**Comment by oskarTeodor** [comment]
For anyone that's stuck waiting for the patch I have a solution for local testing that works for me.

1. Download the latest version of Cypress via Direct download : https://docs.cypress.io/guides/getting-started/installing-cypress.html#Direct-download
2. Install and place the Cypress folder at the location of your older versions of Cypress for me it's in: C:\Users\oskar\AppData\Local\Cypress\Cache\3.2.0 (Important; you have to have the same folder structure --> 3.2.0\Cypress\{content of cypress install}
3. Change the package.json in your project to match Cypress version 3.2.0 --> run yarn install (npm install) in your project
4. Start using Cypress as before e.g. yarn cypress run 

I have tried this setup up on two different windows machine and it works on both.

**Comment by IonicOnCouch** [comment]
**1. Download the binary manually here:** https://download.cypress.io/desktop
**2. npm isntall:** CYPRESS_INSTALL_BINARY=/path/to/zipfile/cypress.zip npm install cypress --save-dev

**Comment by datdoan** [comment]
> **1. Download the binary manually here:** https://download.cypress.io/desktop
> **2. npm isntall:** CYPRESS_INSTALL_BINARY=/path/to/zipfile/cypress.zip npm install cypress --save-dev

For windows:
**SET CYPRESS_INSTALL_BINARY**=/path/to/zipfile/cypress.zip npm install cypress

**Comment by HackPoint** [comment]
@datdoan how it is done in MAC, still unable to install cypress?

**Comment by datdoan** [comment]
Hi @HackPoint,

For Mac follow @IonicOnCouch steps below: 

> **1. Download the binary manually here:** https://download.cypress.io/desktop
> **2. npm isntall:** CYPRESS_INSTALL_BINARY=/path/to/zipfile/cypress.zip npm install cypress --save-dev

**Comment by sidcyrus** [comment]
> 
> 
> > **1. Download the binary manually here:** https://download.cypress.io/desktop
> > **2. npm isntall:** CYPRESS_INSTALL_BINARY=/path/to/zipfile/cypress.zip npm install cypress --save-dev
> 
> For windows:
> **SET CYPRESS_INSTALL_BINARY**=/path/to/zipfile/cypress.zip npm install cypress

Hello, can someone please give more details on how to set this ? Is it env variable ? full detailed step by step guide please. I am new to this. thank you

**Comment by IonicOnCouch** [comment]
> > > **1. Download the binary manually here:** https://download.cypress.io/desktop
> > > **2. npm isntall:** CYPRESS_INSTALL_BINARY=/path/to/zipfile/cypress.zip npm install cypress --save-dev
> > 
> > 
> > For windows:
> > **SET CYPRESS_INSTALL_BINARY**=/path/to/zipfile/cypress.zip npm install cypress
> 
> Hello, can someone please give more details on how to set this ? Is it env variable ? full detailed step by step guide please. I am new to this. thank you

Hi sidcyrus,
1. Download the binary manually here: https://download.cypress.io/desktop
2. Open a Terminal-Window
3. Navigate to your project-directory
4. Remove the folder "node_modules" in your project-directory if it's exists
5. Remove the file "package-lock.json" in your project-directory if it's exists
6. Run these command in your Terminal: CYPRESS_INSTALL_BINARY=/path/to/zipfile/cypress.zip npm install cypress --save-dev

**Comment by sidcyrus** [comment]
> 
> 
> > > > **1. Download the binary manually here:** https://download.cypress.io/desktop
> > > > **2. npm isntall:** CYPRESS_INSTALL_BINARY=/path/to/zipfile/cypress.zip npm install cypress --save-dev
> > > 
> > > 
> > > For windows:
> > > **SET CYPRESS_INSTALL_BINARY**=/path/to/zipfile/cypress.zip npm install cypress
> > 
> > 
> > Hello, can someone please give more details on how to set this ? Is it env variable ? full detailed step by step guide please. I am new to this. thank you
> 
> Hi sidcyrus,
> 
>     1. Download the binary manually here: https://download.cypress.io/desktop
> 
>     2. Open a Terminal-Window
> 
>     3. Navigate to your project-directory
> 
>     4. Remove the folder "node_modules" in your project-directory if it's exists
> 
>     5. Remove the file "package-lock.json" in your project-directory if it's exists
> 
>     6. Run these command in your Terminal: `CYPRESS_INSTALL_BINARY=/path/to/zipfile/cypress.zip npm install cypress --save-dev`

Thanks But I still get errror.
CYPRESS_INSTALL_BINARY=/path/to/zipfile/cypress.zip : The term 
'CYPRESS_INSTALL_BINARY=/path/to/zipfile/cypress.zip' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the  path is correct and try again.

I even tried giving path of the folder where cypress.zip is to replace the '/path/to/zipfile/' in above command. I am in the directory where I want to install cypress.
Thank you for your help.

**Comment by jennifer-shehane** [comment]
This was released in [3.3.0](https://on.cypress.io/changelog#3-3-0). This issue will be closed to further comment as the exact issue here was resolved and tested. 

If you're experiencing a bug similar to this in Cypress, please [open a new issue](https://github.com/cypress-io/cypress/issues/new) with a fully reproducible example that we can run. There may be a specific edge case with the issue that we need more detail to fix.
